### PR TITLE
Feat/annotation visibility

### PR DESCRIPTION
**Pull request by ladeirarodolfo**
It includes
- Annotation Visibility feat

IMPORTANT
- Opened a new PR because the old one has a totally wrong branch name.
- I have done compared to other PR:
- Covered the changes we discussed on chat
- Added documentation to some methods and also md files
- Added changes to Freehand tool

### CODE REVIEW

File: packages/tools/src/utilities/planar/filterAnnotationsForDisplay.ts [hunk 1]
```diff
@@ -34,6 +34,10 @@ export default function filterAnnotationsForDisplay(
     // and we don't have a proper way to check distance either since a stack can be
     // composed of multiple unrelated images
     return annotations.filter((annotation) => {
+      if (!annotation.isVisible) {
```
Comment (ladeirarodolfo) [inline packages/tools/src/utilities/planar/filterAnnotationsForDisplay.ts:37]: @JamesAPetts  is this accurate? as far as planarFreehand is concerned!

File: packages/tools/src/utilities/planar/filterAnnotationsWithinSlice.ts [hunk 2]
```diff
@@ -41,6 +41,9 @@ export default function filterAnnotationsWithinSlice(
     const data = annotation.data;
     const point = data.handles.points[0];
 
+    if (!annotation.isVisible) {
```
Comment (ladeirarodolfo) [inline packages/tools/src/utilities/planar/filterAnnotationsWithinSlice.ts:44]: @JamesAPetts  is this accurate? as far as planarFreehand is concerned!

File: packages/tools/src/stateManagement/annotation/FrameOfReferenceSpecificAnnotationManager.ts [hunk 3]
```diff
@@ -179,6 +180,7 @@ export default class FrameOfReferenceSpecificAnnotationManager {
 
     toolSpecificAnnotations.push(annotation);
     checkAndDefineIsLockedProperty(annotation);
+    checkAndDefineIsVisibleProperty(annotation);
```
Comment (ladeirarodolfo) [inline packages/tools/src/stateManagement/annotation/FrameOfReferenceSpecificAnnotationManager.ts:183]: This (and also locked property line above) causes a code smell. I.e there is an inappropriate intimacy between AnnotationManger and Each Tool, but they should be actually totally decoupled.

My recommendation is to:
base/AnnotationTool be responsible decorate the annotation (on addnew annotation method)

Each Tool to invoke its super method to allow this decoration.

File: packages/tools/examples/annotationSelectionAndLocking/index.ts [hunk 4]
```diff
@@ -147,7 +147,7 @@ async function run() {
       },
     ],
   });
-  toolGroup.setToolPassive(BidirectionalTool.toolName);
+  toolGroup.setToolPassive(RectangleROITool.toolName);
```
Comment (sedghi) [inline packages/tools/examples/annotationSelectionAndLocking/index.ts:150]: Curious of why you changed bidirectional to rectangleROI?

File: packages/tools/src/stateManagement/annotation/annotationState.ts [hunk 5]
```diff
@@ -123,7 +123,7 @@ function removeAnnotation(
  * Get the Annotation object by its UID
  * @param annotationUID - The unique identifier of the annotation.
  * @param element - The element that the tool is being used on.
- * @returns A Annotation object.
+ * @returns A Annotation object or undefined if that does not exist.
```
Comment (sedghi) [inline packages/tools/src/stateManagement/annotation/annotationState.ts:126]: I think the return type is also incorrect (has happend before this PR)

File: packages/tools/src/tools/annotation/LengthTool.ts [hunk 6]
```diff
@@ -568,6 +570,10 @@ class LengthTool extends AnnotationTool {
 
       let activeHandleCanvasCoords;
 
+      if (!isAnnotationVisible(annotationUID)) {
```
Comment (sedghi) [inline packages/tools/src/tools/annotation/LengthTool.ts:573]: put it after cachedStats

File: packages/tools/src/types/AnnotationTypes.ts [hunk 7]
```diff
@@ -7,6 +7,8 @@ type Annotation = {
   highlighted?: boolean;
   /** If the annotation is locked for manipulation */
   isLocked?: boolean;
+  /** If the annotation is visible for manipulation */
+  isVisible?: boolean;
```
Comment (sedghi) [inline packages/tools/src/types/AnnotationTypes.ts:11]: Why this is optional? I thought it is always there and default to true

File: packages/docs/docs/concepts/cornerstone-tools/annotation/visibility.md [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+import { annotations } from '@cornerstonejs/tools';
+
+// changing an annotation visibility to be visible (implicit visible param).
+annotations.visibility.setAnnotationVisibility(annotationUID);
+
+// changing an annotation visibility to NOT be visible.
+annotations.visibility.setAnnotationVisibility(annotationUID, false);
+
+// get all the hidden annotations uids.
+annotations.visibility.getAnnotationUIDsHidden();
```
Comment (sedghi) [inline packages/docs/docs/concepts/cornerstone-tools/annotation/visibility.md:25]: hmm, not sure if it is me only, but It feels weird that the API has both `visible` and `hidden` methods. Can't we say `getAnnotationUIDsInvisible`? or `getAnnotationUIDsInvisibleCount`? This might be only me since I'm not well known for naming things :D @JamesAPetts

File: packages/tools/src/stateManagement/annotation/FrameOfReferenceSpecificAnnotationManager.ts (hunk 3 above)
Comment (ladeirarodolfo) [reply packages/tools/src/stateManagement/annotation/FrameOfReferenceSpecificAnnotationManager.ts:183]: ok
https://github.com/cornerstonejs/cornerstone3D-beta/issues/95

File: packages/tools/examples/annotationSelectionAndLocking/index.ts (hunk 4 above)
Comment (ladeirarodolfo) [reply packages/tools/examples/annotationSelectionAndLocking/index.ts:150]: I shouldnt commit that at all.. it was just because of testing purpose

File: packages/tools/src/stateManagement/annotation/annotationState.ts (hunk 5 above)
Comment (ladeirarodolfo) [reply packages/tools/src/stateManagement/annotation/annotationState.ts:126]: Glad that I forgot that.
I will remove this, but I would like to do an observation that the signature of it is not consistent to
annotationManager.getAnnotation(annotationUID);

File: packages/tools/src/types/AnnotationTypes.ts (hunk 7 above)
Comment (ladeirarodolfo) [reply packages/tools/src/types/AnnotationTypes.ts:11]: This is indeed a very trick one. It tells us something about our architecture: 
As far as OOO is concerned, this type is not what it meant to be. 
When I see this type Annotation, I understand this represents an instance of an AnnotationTool. If that is true, each annotation instance has at least to 2 phases/shapes types. I.e at construction time and after adding the handles.
So, with all that said that is why many of flags here are optional, because they are completed (if that happens) at run time. 
In summary, AnnotationType is being used as a representation of instance (at construction time) and considering this flag is added by decoration at run time (not at construction time), I believe would be fine to let it optional. This should be fixed whenever we move the decoration to BaseAnnotationLayer instead of AnnotationManager layer

File: packages/tools/src/stateManagement/annotation/annotationVisibility.ts [hunk 9]
```diff
@@ -73,7 +73,7 @@ function isAnnotationVisible(annotationUID: string): boolean | undefined {
 /**
  * Get the number of hidden annotation objects in the global set of hidden annotation
  * objects uids.
- * @returns {number} The number of hidden annotation objects.
+ * @returns The number of hidden annotation objects.
  *
  */
 function getAnnotationUIDsHiddenCount(): number {
```
Comment (sedghi) [inline packages/tools/src/stateManagement/annotation/annotationVisibility.ts:79]: Also this doesn't make sense it has UID in the name, can we have `getNumberOfHiddenAnnotations` and any one call .length on it

File: packages/docs/docs/concepts/cornerstone-tools/annotation/visibility.md (hunk 8 above)
Comment (ladeirarodolfo) [reply packages/docs/docs/concepts/cornerstone-tools/annotation/visibility.md:25]: This is not a requirement on my impl. So I will remove it. 
If there is a need for something like that in future we can add it back.

File: packages/tools/src/stateManagement/annotation/annotationVisibility.ts [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+import { AnnotationVisibilityChangeEventDetail } from '../../types/EventTypes';
+
+/*
+ * It stores all hidden annotation uids.
+ */
+const globalHiddenAnnotationUIDsSet: Set<string> = new Set();
+
+/*
+ * Interface (Public API)
+ */
```
[the remark this answers is not in the document]
Comment (JamesAPetts) [reply packages/tools/src/stateManagement/annotation/annotationVisibility.ts:14]: Its certainly some pretty strange code on both fronts (not criticising this PR, it was obviously following the same pattern):

`checkAndDefineIsVisibleProperty` defines the getter and setter of the property but it does seem like they aren't used. There doesn't seem to be anywhere we are setting isVisible directly anywhere, but it also looks like if you traverse the `setAnnotationVisibility` tree you don't hit the defined properties. But we _are_ checking the properties of the annotations, so they must be being set somehow. We should clean up this implementation so its less esoteric at some point, it really needn't be this complicated.

File: packages/tools/src/types/AnnotationTypes.ts (hunk 7 above)
Comment (JamesAPetts) [reply packages/tools/src/types/AnnotationTypes.ts:11]: >When I see this type Annotation, I understand this represents an instance of an AnnotationTool

This is incorrect. The`Annotation` is _data_ created by an `AnnotationTool`, not an instance of the `AnnotationTool`. `Annotations` are like `toolData` in Cornerstone Legacy.

`isVisible` should be mandatory, and set true by default when the annotation is defined in an `AnnotationTool`'s `addNewAnnotation` method (or as you said before,as a super called to `BaseAnnotationTool`).

There should be no scenario where `isVisible` is undefined, and we should really do the same for `isLocked` too (set it `false` on creation).

cc @sedghi .

File: packages/tools/src/utilities/planar/filterAnnotationsWithinSlice.ts (hunk 2 above)
Comment (JamesAPetts) [reply packages/tools/src/utilities/planar/filterAnnotationsWithinSlice.ts:44]: Same as above.

**Review by JamesAPetts — CHANGES_REQUESTED** [review]
Some minor changes to example required, the rest are discussion points, and don't necessarily need to be dealt with in this PR.

**Review by JamesAPetts — APPROVED** [review]
Looks good for now, we can make the proposed improvements to base annotation tool etc in the future, this PR is good to merge.
